src/services/dashboards/dashboard-customize/modules/VariableMoreButtonDropdown.vue

```diff
@@ -45,64 +54,99 @@ interface Props {
     variableMap: DashboardVariablesSchema['properties'];
     variableOrder: string[];
 }
-
 interface EventEmits {
     (e: string, value: string): void;
-    (e: 'manage-more'): void;
-    (e: 'select', value: DashboardVariablesSchema['properties']): void;
+    (e: 'change', value: DashboardVariablesSchema['properties']): void;
 }
 
 const props = defineProps<Props>();
 const emit = defineEmits<EventEmits>();
 
 const state = reactive({
-    visibleMenu: false,
-    variables: computed(() => {
-        const properties = props.variableMap;
-        const menuItems = [] as any[];
-        Object.keys(properties).forEach((d) => menuItems.push({
-            name: d,
-            label: properties[d].name,
-        }));
-        return menuItems;
-    }),
-    order: computed(() => props.variableOrder),
+    targetRef: null,
+    contextMenuRef: null,
+    variables: [] as MenuItem[],
     selected: [] as MenuItem[],
 });
 
+const {
+    targetRef,
+    contextMenuRef,
+    variables,
+    selected,
+} = toRefs(state);
+
+const {
+    visibleMenu,
+    hideContextMenu,
+    showContextMenu,
+    fixedMenuStyle,
+    reorderMenuBySelection,
+    menu,
+} = useContextMenuController({
+    targetRef,
+    contextMenuRef,
+    useReorderBySelection: true,
+    useFixedStyle: true,
+    menu: variables,
+});
+
 // event
 const handleOpenOverlay = () => {
     SpaceRouter.router.push({ hash: MANAGE_VARIABLES_HASH_NAME });
 };
 const handleClickButton = () => {
-    state.visibleMenu = !state.visibleMenu;
-};
-const handleVariableSelect = (item: MenuItem) => {
-    if (!item.name) return;
-    const result = cloneDeep(props.variableMap) as DashboardVariablesSchema['properties'];
-    if (props.variableMap[item.name]) {
-        result[item.name].use = !result[item.name].use;
+    if (visibleMenu.value) {
+        hideContextMenu();
+    } else {
+        reorderMenuBySelection(selected.value);
+        showContextMenu();
     }
-    emit('select', result);
 };
 
+// After context menu is hidden, update selected variable's use.
+watch(() => visibleMenu.value, () => {
+    if (visibleMenu.value) return;
+    const properties = cloneDeep(props.variableMap) as DashboardVariablesSchema['properties'];
+    props.variableOrder.forEach((d) => {
+        properties[d].use = state.selected.some((item) => item.name === d);
+    });
+    emit('change', properties);
+});
```

Comment:
```suggestion
watch([visibleMenu, () => state.selected], (after, before) => {
   if (!after) return;
    const [_visibleMenu, selected] = after
    if (_visibleMenu) return;
    const [, beforeSelected] = before
    if (selected === beforeSelected) return;
    const properties = cloneDeep(props.variableMap) as DashboardVariablesSchema['properties'];
    props.variableOrder.forEach((d) => {
        properties[d].use = selected.some((item) => item.name === d);
    });
    emit('change', properties);
});
```

Reply:
issue: I think must calculate order of variables based on selected and emit variableOrder too.

```suggestion
watch([visibleMenu, () => state.selected], (after, before) => {
   if (!after) return;
    const [_visibleMenu, selected] = after
    if (_visibleMenu) return;
    const [, beforeSelected] = before
    if (selected === beforeSelected) return;

    const properties = cloneDeep(props.variableMap) as DashboardVariablesSchema['properties'];
    const variableOrder = selected.map(item => {
        properties[item.name].use = true;
        return item.name;
    })
    emit('change', properties, variableOrder);
});
```